cmd{containerboot,k8s-operator},util/linuxfw: support ExternalName Services

irbekrm opened the pull request:
See https://github.com/tailscale/tailscale/issues/10606 for context, especially  https://github.com/tailscale/tailscale/issues/10606#issuecomment-1858858609

This PR adds initial support to exposing targets accessible from within the Kubernetes cluster to tailnet via DNS names using an [`ExternalName` `Service`](https://kubernetes.io/docs/concepts/services-networking/service/#externalname).

To expose a target accessible from within a cluster:

## Flow

1. A user would create an `ExternalName` `Service`:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: foo 
  annotations:
    tailscale.com/expose: "true"
spec:
  type: ExternalName
  externalName: <dns-name>
```

2. The Kubernetes operator will reconcile the `Service` and create a proxy based on containerboot  deployment with [`TS_DEST_DNS_NAME` env var](https://github.com/tailscale/tailscale/pull/11802/commits/dff1eeb47257e1336bde1fd1898132190ac54668#diff-700830ac79a41da42cd3a82065f0fee3003a4390eead4db1de4bf30557a06bd6R22) set to the DNS name from the `ExternalName` `Service`

3. The proxy will resolve the DNS name once on start and then every 10 minutes. It will ensure that traffic received on the proxy's tailnet IP address is forwarded to the backends.

4. Users should then be able to reach the backends from the tailnet via the tailscale IP address or MagicDNS name of the proxy, i.e
```
$ kubectl get secret -n tailscale -l tailscale.com/managed="true",tailscale.com/parent-resource=foo,tailscale.com/parent-resource-ns=default -ojson | jq -r '.items[0].data.device_ips' | base64 -d
["100.44.55.66", "fd7a:115c:a1e0::8e01:2765"]
$ curl -L 100.44.55.66
...
```

## Notes

- for target DNS names that resolve to multiple IP addresses the traffic is only load balanced accross alll addresses when the firewall mode is iptables. When the firewall mode is nftables, the traffic is sent to the first IP address. This needs fixing, I've tried to set up load balancing with nftables, but currently am seeing weird issues for repeated connections see [here](https://github.com/tailscale/tailscale/commit/d37f2f508509c6c35ad724fd75a27685b90b575b#diff-a3bcbcd1ca198799f4f768dc56fea913e1945a6b3ec9dbec89325a84a19a85e7R148-R232).

oxtoacart commented:
I was able to see that this works in my own environment, will review code now.

oxtoacart commented on cmd/containerboot/main.go:
```diff
@@ -441,12 +459,32 @@ runLoop:
+				if cfg.ProxyTargetDNSName != "" {
+					newBackendAddrs, err := resolveDNS(ctx, cfg.ProxyTargetDNSName)
+					if err != nil {
+						log.Printf("unable to resolve DNS name %s: %v, retrying in %s", cfg.ProxyTargetDNSName, err, proxyTargetIPsResolvePeriod)
```
If we're unlucky and happen to hit a transient network issue on our first attempt to resolveDNS, we'll end up having to wait for 10 minutes before this starts working. It might be better to retry with an exponential backoff when we encounter an error.

Instead of doing this on a static ticker, what if we did something like this:

1. On initialization, perform the lookup and firewall configuration right away
2. Then, on every connection request, perform the DNS lookup again. If the IPs changed, reconfigure the firewall.

Yet another option would be to look at the TTL of the DNS responses and schedule the recheck/reconfigure based on that.

irbekrm replied:
> Then, on every connection request, perform the DNS lookup again

We don't have anything in code that is aware of connection requests and I am not sure whether the connection request/IP address change patterns would be such that this would actually be good user experience.

We _could_ reduce the period till the next lookup on a failure with an exponential backoff for repeated lookups post failures up till whatever is the static period. And yeah could replace the static 10 minutes with TTL 👍🏼

oxtoacart replied:
> We don't have anything in code that is aware of connection requests

Oh yeah, good point!

> We could reduce the period till the next lookup on a failure with an exponential backoff for repeated lookups post failures up till whatever is the static period. And yeah could replace the static 10 minutes with TTL

Sounds good.

irbekrm replied:
@oxtoacart I ended up not using TTL - that would have needed pulliing in miekg/dns and I don't want to (yet) parse in the dns responses from there to extract TTL. I was playing around with our internal recursive.Resolver for this that returns a more suitable response, but that seems to return very unexpectedly short TTLs- I want to look more at what is the expected behaviour from that, but I think for this PR keeping the default and the max check period to 10 minutes could be good enough- we can see how well it works for users who want this.

I've added logic to start with shorter checks (from 2s to 10 minutes, exponentially upwards) for consecutive failing resolution checks.

irbekrm replied:
https://github.com/tailscale/tailscale/pull/11802/commits/e2a1b15777208754ef5e35561ef6d24c246f977f#diff-700830ac79a41da42cd3a82065f0fee3003a4390eead4db1de4bf30557a06bd6R390-R411

oxtoacart commented on cmd/containerboot/main.go:
```diff
@@ -918,15 +995,79 @@ func installIngressForwardingRule(ctx context.Context, dstStr string, tsIPs []ne
+		if err := nfr.DNATWithLoadBalancer(dst, backendTargets); err != nil {
+			return fmt.Errorf("installing DNAT rules for ingress backends %+#v: %w", backendTargets, err)
+		}
+		// The backend might advertize MSS higher than that of the
```
Very nice! I once lost days to this exact problem, so I really appreciate this.

irbekrm replied:
yeah we do it for all the kube proxies and in some Kubernetes environments it seems like PMTU discovery is actually not working see i.e https://github.com/tailscale/tailscale/issues/11002

oxtoacart commented on cmd/containerboot/main.go:
```diff
@@ -918,15 +995,79 @@ func installIngressForwardingRule(ctx context.Context, dstStr string, tsIPs []ne
 	// is done. This is typically a locally reachable IP.
-	ProxyTo string
+	ProxyTargetIP string
+	// ProxyTargetDNSName is a DNS name whose backing IP addresses all
```
nit: ProxyTargetDNSName is a DNS name **to** whose backing IP addresses all incoming Tailscale traffic should be proxied .

oxtoacart commented on cmd/containerboot/main.go:
```diff
@@ -993,6 +1140,16 @@ func (s *settings) validate() error {
 }
 
+func resolveDNS(ctx context.Context, name string) ([]net.IP, error) {
+	ips, err := net.LookupIP(name)
```
Consider using `net.DefaultResolver.LookupIP(ctx, name)` to respect any deadlines on ctx.

oxtoacart commented on cmd/containerboot/main.go:
```diff
@@ -772,6 +829,26 @@ func ensureIPForwarding(root, clusterProxyTarget, tailnetTargetiP, tailnetTarget
 			v6Forwarding = true
 		}
 	}
+	if clusterProxyTargetDNSName != "" {
```
Consider using `else if` since `clusterProxyTargetIP` and `clusterProxyTargetDNSName` are mutually exclusive.

irbekrm replied:
We would fail at [validation if both were specified](https://github.com/tailscale/tailscale/pull/11802/commits/477b8d0483b6caec5a8998791f92b363e7ace1ab#diff-700830ac79a41da42cd3a82065f0fee3003a4390eead4db1de4bf30557a06bd6R1116)

oxtoacart replied:
Yes, I saw the validation, just thought it would be nice to reinforce that these are mutually exclusive in this branching logic. Not a big deal either way.

irbekrm replied:
I've left this as is for now - I would find the code blocks harder to read myself with even more indentation. We do need to refactor the containerboot code as it is growing out quite wildly now, there is an idea to implement a TCP proxy in code

oxtoacart replied:
Totally cool. FWIW, using `else if` wouldn't add any indentation :)

irbekrm replied:
more characters :) `if` -> `else if`

oxtoacart commented on cmd/containerboot/main.go:
```diff
@@ -150,8 +155,8 @@ func main() {
-		if cfg.ProxyTo != "" || cfg.Routes != nil || cfg.TailnetTargetIP != "" || cfg.TailnetTargetFQDN != "" {
-			if err := ensureIPForwarding(cfg.Root, cfg.ProxyTo, cfg.TailnetTargetIP, cfg.TailnetTargetFQDN, cfg.Routes); err != nil {
+		if cfg.ProxyTargetIP != "" || cfg.ProxyTargetDNSName != "" || cfg.Routes != nil || cfg.TailnetTargetIP != "" || cfg.TailnetTargetFQDN != "" {
+			if err := ensureIPForwarding(cfg.Root, cfg.ProxyTargetIP, cfg.ProxyTargetDNSName, cfg.TailnetTargetIP, cfg.TailnetTargetFQDN, cfg.Routes); err != nil {
```
There's a race condition here. `ensureIPForwarding()` only runs once. It does a DNS lookup to determine whether IPv4 or IPv6 forwarding are needed. Subsequently, we perform periodic DNS lookups down in the `runLoop`. In theory, those could return IPv6 when the initial lookup didn't, at which point we may not have the right forwarding enabled.

It might be better to call `ensureIPForwarding` whenever the resolved addrs change.

irbekrm replied:
That's a good catch, I think we need to do the same for other proxy types too, so I'll do this in a separate PR

oxtoacart replied:
I think(?) the problem is unique to the new DNS-based flow because of the DNS lookup. For static IPs, we know up-front whether it's IPv4 or IPv6.

irbekrm replied:
Yeah I was thinking tailnet IPs of the node that can change during the proxy Pod's run- however there is actually no case that I can think of where adding an IPv4/IPv6 tailnet address to a node that previously did not have it would result in reconfiguring firewall for an already running proxy Pod.

I've split the `ensureIPForwarding` and am only running it at the point when we know we need to configure firewall to avoid failing DNS resolution crashing containerboot upfront.
In practice, since this is mostly for kube, we likely don't need this at all because of https://github.com/tailscale/tailscale/blob/main/cmd/k8s-operator/deploy/manifests/proxy.yaml#L20

oxtoacart reviewed (commented):
Very cool! Most of my questions/concerns are about the timing of DNS lookups and reconfiguring the firewall.

irbekrm commented on cmd/containerboot/main.go:
```diff
@@ -362,6 +370,16 @@ authLoop:
+	var ts time.Ticker
+	if cfg.ProxyTargetDNSName != "" {
+		ts = *time.NewTicker(proxyTargetIPsResolvePeriod)
+	}
```
👍🏼 (it runs in a container though so all resources will be thrown away on stop anyway)

irbekrm commented on cmd/containerboot/main.go:
```diff
@@ -441,12 +459,32 @@ runLoop:
+						return -1
+					})
+					if len(addrs) > 0 && (backendsHaveChanged != 0 || ipsHaveChanged) && len(newBackendAddrs) > 0 {
+						log.Printf("installing ingresss proxy rules for backends %v", newBackendAddrs)
```
TIL

irbekrm commented on cmd/containerboot/main.go:
```diff
@@ -441,12 +459,32 @@ runLoop:
+						log.Printf("unable to resolve DNS name %s: %v, retrying in %s", cfg.ProxyTargetDNSName, err, proxyTargetIPsResolvePeriod)
+						continue
+					}
+					backendsHaveChanged := slices.CompareFunc(backendAddrs, newBackendAddrs, func(ip1 net.IP, ip2 net.IP) int {
```
TIL

irbekrm commented on cmd/containerboot/main.go:
```diff
@@ -441,12 +459,32 @@ runLoop:
+					if len(addrs) > 0 && (backendsHaveChanged != 0 || ipsHaveChanged) && len(newBackendAddrs) > 0 {
+						log.Printf("installing ingresss proxy rules for backends %v", newBackendAddrs)
+						if err := installIngressForwardingRuleExternalNameService(ctx, newBackendAddrs, addrs, nfr); err != nil {
+							log.Fatalf("error installing ingress proxy rules: %v", err)
```
It is yeah, this is really only expected to run in (kube) containers, where if the proxy setup fails we should crash and let kube restart the container + the user to notice that something is wrong

irbekrm commented:
> Very cool! Most of my questions/concerns are about the timing of DNS lookups and reconfiguring the firewall.

Thanks a lot for the thorough review Percy, I believe I have addressed all the comments now.

oxtoacart commented on cmd/containerboot/main.go:
```diff
@@ -362,6 +373,44 @@ authLoop:
+
+	// Setup for proxies that are configured to proxy to a target specified
+	// by a DNS name (TS_EXPERIMENTAL_DEST_DNS_NAME).
+	const defaultCheckPeriod = time.Minute * 10 // how often to check if what IPs the DNS name resolves to
```
nit: "how often to check what IPs the DNS name resolves to"

oxtoacart replied:
nit: "how often to check what IPs the DNS name resolves to"

oxtoacart commented on cmd/containerboot/main.go:
```diff
@@ -511,12 +579,29 @@ runLoop:
+			backendsHaveChanged := !(slices.EqualFunc(backendAddrs, newBackendAddrs, func(ip1 net.IP, ip2 net.IP) bool {
+				return slices.ContainsFunc(newBackendAddrs, func(ip net.IP) bool { return ip.Equal(ip1) })
+			}))
+			if backendsHaveChanged && len(addrs) != 0 {
```
nit: on line 499, you use `len(addrs) > 0` and here you use `len(addrs) != 0`

irbekrm replied:
Thanks, I've changed all `> 0`s to `!= 0`

oxtoacart commented on cmd/containerboot/main.go:
```diff
@@ -757,12 +842,12 @@ func ensureTunFile(root string) error {
 
 // ensureIPForwarding enables IPv4/IPv6 forwarding for the container.
-func ensureIPForwarding(root, clusterProxyTarget, tailnetTargetiP, tailnetTargetFQDN string, routes *string) error {
+func ensureIPForwarding(root, clusterProxyTargetIP, tailnetTargetiP, tailnetTargetFQDN string, routes *string) error {
```
nit: tailnetTargetiP -> tailnetTargetIP

oxtoacart reviewed (approved):
LGTM! Just noted a few nits, nothing impactful.
